specs/AdditionalAllowedFrameAncestors.md

```diff
@@ -0,0 +1,234 @@
+}
+
+// App specific logic to decide whether a site is the one it wants to host.
+bool IsTargetSite(PCWSTR siteUrl)
+{
+    return AreSitesSame(siteUrl, siteToHost);
+}
+
+void MyApp::HandleHostedSites()
+{
+    CHECK_FAILURE(m_webview->add_FrameCreated(
+        Callback<ICoreWebView2FrameCreatedEventHandler>(
+            [this](ICoreWebView2* sender, ICoreWebView2FrameCreatedEventArgs* args)
+                -> HRESULT 
+            {
+                wil::unique_cotaskmem_string pageUrl;
+                CHECK_FAILURE(m_webView->get_Source(&pageUrl));
+                // IsAppContentUri verifies that pageUrl is app's content.
+                if (IsAppContentUri(pageUrl.get()))
+                {
+                    // We are on trusted pages. Now check whether it is the iframe we plan
+                    // to host other sites.
+                    const std::wstring siteHostingFrameName = L"my_site_hosting_frame";
+                    wil::com_ptr<ICoreWebView2Frame> webviewFrame;
+                    CHECK_FAILURE(args->get_Frame(&webviewFrame));
+                    wil::unique_cotaskmem_string frameName;
+                    CHECK_FAILURE(webviewFrame->get_Name(&frameName));
+                    if (siteHostingFrameName == frameName.get())
+                    {
+                      // We are hosting sites.
+                      m_hostingSite = true;
+                      CHECK_FAILURE(webviewFrame->add_Destroyed(
+                          Microsoft::WRL::Callback<
+                              ICoreWebView2FrameDestroyedEventHandler>(
+                              [this](ICoreWebView2Frame* sender,
+                                     IUnknown* args) -> HRESULT {
+                                m_hostingSite = false;
+                                return S_OK;
+                              })
+                              .Get(),
+                          nullptr));
+                    }
+                }
+                return S_OK;
+            })
+            .Get(),
+        nullptr));
+    // Using FrameNavigationStarting event instead of NavigationStarting event of CoreWebViewFrame
+    // to cover all possible nested iframes inside the hosted site as CoreWebViewFrame
+    // object currently only support first level iframes in the top page.
+    CHECK_FAILURE(m_webview->add_FrameNavigationStarting(
+        Microsoft::WRL::Callback<ICoreWebView2NavigationStartingEventHandler>(
+            [this](
+                ICoreWebView2* sender,
+                ICoreWebView2NavigationStartingEventArgs* args) -> HRESULT
+            {
+              if (m_hostingSite)
+              {
+                  wil::unique_cotaskmem_string navigationTargetUri;
+                  CHECK_FAILURE(args->get_Uri(&navigationTargetUri));
+                  wil::com_ptr<
+                      ICoreWebView2NavigationStartingEventArgs_2>
+                      navigationStartArgs;
+                  if (SUCCEEDED(args->QueryInterface(
```

Comment:
Confusing. Should do IsTargetSite first.

Reply:
If the app is not expected to embed any site at the moment, it should not do anything. Also checking a bool m_hostingSite is much less costly.

Reply:
Can swap or separate the navigationTargetUri check and QI to make it easier to read. Please do.